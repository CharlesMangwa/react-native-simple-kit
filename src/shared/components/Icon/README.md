# Icon ðŸ’ 

A generic component for custom SVG icons.

### API

- name: `'backButton' | 'home' | 'people' | 'profile'`
- activated?: `boolean`
- size?: `number`
- defaultActivatedColor?: `string`
- defaultColor?: `string`
- style?: `StyleSheet`

### Example

```js
import React from 'react'
import Icon from '@components/Icon'

const MyComponent = (): React$Element<*> => (
  <Icon name="backButton" defaultColor="red" size={22} />
)

export default MyComponent
```

## Changelog

#### Sunday, June 24th

- Add `home`, `people` & `profile` icons.

#### Thursday, November 23th

- Rename `actived` prop into `activated`.
- Rename `defaultactivedColor` prop into `defaultActivatedColor`.

#### Wednesday, November 15th

- Initial release. We didn't track changes before this version.
