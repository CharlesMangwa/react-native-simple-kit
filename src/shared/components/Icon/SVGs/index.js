/* @flow */

export { default as Back } from './Back'
export { default as Home } from './Home'
export { default as People } from './People'
export { default as Profile } from './Profile'
