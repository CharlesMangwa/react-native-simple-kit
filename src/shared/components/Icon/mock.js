/* eslint global-require: 0 */

jest.mock('react-native-svg', () => {
  const React = require('react')
  // eslint-disable-next-line no-undef
  const ReactNativeSvg = jest.genMockFromModule('react-native-svg')

  const svgElementMockGenerator = (name, propTypes) => {
    const SvgMock = () => React.createElement(name, null, null)

    SvgMock.displayName = name
    SvgMock.propTypes = propTypes

    return SvgMock
  }

  const Svg = svgElementMockGenerator('Svg', ReactNativeSvg.Svg.propTypes)

  Svg.Circle = svgElementMockGenerator(
    'Circle',
    ReactNativeSvg.Circle.propTypes
  )
  Svg.Defs = svgElementMockGenerator('Defs', ReactNativeSvg.Defs.propTypes)
  Svg.G = svgElementMockGenerator('G', ReactNativeSvg.G.propTypes)
  Svg.Path = svgElementMockGenerator('Path', ReactNativeSvg.Path.propTypes)
  Svg.Rect = svgElementMockGenerator('Rect', ReactNativeSvg.Rect.propTypes)
  Svg.Stop = svgElementMockGenerator('Stop', ReactNativeSvg.Stop.propTypes)
  Svg.Symbol = svgElementMockGenerator(
    'Symbol',
    ReactNativeSvg.Symbol.propTypes
  )
  Svg.Use = svgElementMockGenerator('Use', ReactNativeSvg.Use.propTypes)

  return Svg
})
