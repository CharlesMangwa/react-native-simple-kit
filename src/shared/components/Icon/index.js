/* @flow */

import React, { createElement } from 'react'
import PropTypes from 'prop-types'
import { View } from 'react-native'

import { NEUTRAL_COLOR_50, BRAND_COLOR_GREEN } from '@theme/colors'
import * as SVGs from './SVGs'

type Props = {
  activated?: boolean,
  defaultActivatedColor?: string,
  defaultColor?: string,
  name: string,
  size: number,
  style?: ?Array<?StyleSheet> | StyleSheet,
}

const Icon = (props: Props): React$Element<*> => {
  const {
    name,
    style,
    size,
    activated,
    defaultColor,
    defaultActivatedColor,
  } = props
  const color = activated
    ? defaultActivatedColor || BRAND_COLOR_GREEN
    : defaultColor || NEUTRAL_COLOR_50

  let icon
  switch (name) {
    case 'backButton':
      icon = SVGs.Back
      break
    case 'home':
      icon = SVGs.Home
      break
    case 'people':
      icon = SVGs.People
      break
    case 'profile':
      icon = SVGs.Profile
      break
    default: {
      icon = createElement(View, { ...props })
      throw new Error(`Invalid SVG name: ${name}`)
    }
  }

  return (
    <View {...props} style={style}>
      {createElement(icon, { color, size })}
    </View>
  )
}

Icon.propTypes = {
  activated: PropTypes.bool,
  defaultActivatedColor: PropTypes.string,
  defaultColor: PropTypes.string,
  name: PropTypes.string.isRequired,
  size: PropTypes.number.isRequired,
  style: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),
}

Icon.defaultProps = {
  activated: false,
  defaultActivatedColor: BRAND_COLOR_GREEN,
  defaultColor: NEUTRAL_COLOR_50,
  style: undefined,
}

export default Icon
