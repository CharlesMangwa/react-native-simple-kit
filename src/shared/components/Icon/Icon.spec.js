import './mock.js'
import React from 'react'
import TestRenderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'
import Icon from './index.js'

describe('ðŸ’  Generic component: <Icon />', () => {
  it('should render correctly', () => {
    const component = TestRenderer.create(<Icon name="backButton" size={15} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('should render activated version correctly', () => {
    const component = TestRenderer.create(
      <Icon activated name="backButton" size={15} />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('should render with activated version with custom color correctly', () => {
    const component = TestRenderer.create(
      <Icon
        activated
        defaultActivatedColor="#5AADEB"
        name="backButton"
        size={15}
      />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('should render with defaultColor correctly', () => {
    const component = TestRenderer.create(
      <Icon defaultColor="#333333" name="backButton" size={15} />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("should throw when the required SVG doesn't exist", () => {
    const renderer = new ShallowRenderer()
    expect(() =>
      renderer.render(<Icon name="throw me an error!" size={15} />)
    ).toThrowError('Invalid SVG name: throw me an error!')
  })
})
