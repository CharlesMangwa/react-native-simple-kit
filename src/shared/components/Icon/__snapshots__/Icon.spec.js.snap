// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ’  Generic component: <Icon /> should render activated version correctly 1`] = `
<View
  activated={true}
  defaultActivatedColor="#53C000"
  defaultColor="#b8b8b8"
  name="backButton"
  size={15}
>
  <Svg />
</View>
`;

exports[`ðŸ’  Generic component: <Icon /> should render correctly 1`] = `
<View
  activated={false}
  defaultActivatedColor="#53C000"
  defaultColor="#b8b8b8"
  name="backButton"
  size={15}
>
  <Svg />
</View>
`;

exports[`ðŸ’  Generic component: <Icon /> should render with activated version with custom color correctly 1`] = `
<View
  activated={true}
  defaultActivatedColor="#5AADEB"
  defaultColor="#b8b8b8"
  name="backButton"
  size={15}
>
  <Svg />
</View>
`;

exports[`ðŸ’  Generic component: <Icon /> should render with defaultColor correctly 1`] = `
<View
  activated={false}
  defaultActivatedColor="#53C000"
  defaultColor="#333333"
  name="backButton"
  size={15}
>
  <Svg />
</View>
`;
